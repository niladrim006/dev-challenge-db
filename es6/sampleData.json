[
    {
        "bestAsk": 1.255268054923685,
        "bestBid": 1.234401281999198,
        "lastChangeAsk": -0.010712388852019084,
        "lastChangeBid": 0.02224912511866406,
        "midPrice": [],
        "name": "gbpeur",
        "openAsk": 1.2924134774154548,
        "openBid": 1.283586522584545
    },
    {
        "bestAsk": 108.5276339286429,
        "bestBid": 105.37527644126496,
        "lastChangeAsk": -0.02883057593133742,
        "lastChangeBid": -0.7686749906997221,
        "midPrice": [
            {
                "timeStamp": 1652379744000,
                "midPrice": 1.3254518885616453
            },
            {
                "timeStamp": 1652379745000,
                "midPrice": 1.2987548278762557
            },
            {
                "timeStamp": 1652379747000,
                "midPrice": 1.2589728869898302
            },
            {
                "timeStamp": 1652379748000,
                "midPrice": 1.252424921400401
            },
            {
                "timeStamp": 1652379751000,
                "midPrice": 1.2376755755816133
            }
        ],
        "name": "usdjpy",
        "openAsk": 110.34002036807428,
        "openBid": 106.66997963192571
    }
]