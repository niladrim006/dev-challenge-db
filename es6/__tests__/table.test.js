import CurrencyTable from '../table';
import sampleData from '../sampleData.json';

describe('Sorted Datatable functionality test', () => {


  it('should receive message, data assignment and class addition on table', () => {
    const message = { body: JSON.stringify(sampleData[0])};
    document.body.innerHTML = `
      <div id=\"table-body\">\n
        <div></div>\n
      </div>
    `;
    window.Sparkline = { draw: jest.fn() }
    const dataTable = new CurrencyTable();

    jest.spyOn(dataTable, 'subscribeData');
    // tests for subscribeData()
    dataTable.subscribeData(message);

    expect(dataTable.subscribeData).toHaveBeenCalledTimes(1);
    expect(dataTable.subscribeData).toHaveBeenCalledWith(message);

    // // ensure data is consistent
    expect(dataTable.tableMapData.size).toBe(1);

    // check for template
    expect(document.getElementById('table-body')).toBeTruthy();
  });

  it('should invoke mapData, sortMap and displaySortedData when recieves message', () => {
    const message = { body: JSON.stringify(sampleData[0])};
    const dataTable = new CurrencyTable();

    jest.spyOn(dataTable, 'mapData').mockImplementation(()=> null);
    jest.spyOn(dataTable, 'sortMap').mockImplementation(()=> null);
    jest.spyOn(dataTable, 'displaySortedData').mockImplementation(()=> null);

    dataTable.subscribeData(message);
    
    // tests for mapData, sortMap, displaySortedData on  subscribeData() runs
    expect(dataTable.mapData).toHaveBeenCalled();
    expect(dataTable.sortMap).toHaveBeenCalled();
    expect(dataTable.displaySortedData).toHaveBeenCalled();
  })

  it('should set data on tableMapData map object when mapData invokes with data for fresh entry', () => {
    const dataTable = new CurrencyTable();
    dataTable.mapData(sampleData[0])
    expect(dataTable.tableMapData.size).toBe(1);
  })

  it('should set data on tableMapData map object when mapData invokes with data existing data', () => {
    const dataTable = new CurrencyTable();
    dataTable.tableMapData.set(sampleData[0].name, sampleData[0])
    dataTable.mapData(sampleData[0])
    expect(dataTable.tableMapData.size).toBe(1);
  })

  it('should short the table data based on lastChangeBid', () => {
    const dataTable = new CurrencyTable();
    dataTable.tableMapData.set(sampleData[0].name, sampleData[0])
    dataTable.tableMapData.set(sampleData[1].name, sampleData[1])

    // test before sorting
    expect([...dataTable.tableMapData.entries()][0][1].lastChangeBid).toBe(0.02224912511866406);
    dataTable.sortMap()
    // test before sorting
    expect([...dataTable.tableMapData.entries()][1][1].lastChangeBid).toBe(0.02224912511866406);
  })
  
  it('should return array of sparkline point by Taking object', () => {
    const dataTable = new CurrencyTable();
    let data = dataTable.getSparklineGraph(sampleData[1].midPrice)
    expect(Array.isArray(data)).toBeTruthy();
  })

});