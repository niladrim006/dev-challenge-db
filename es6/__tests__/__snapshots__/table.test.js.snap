// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sorted Datatable functionality test should receive message, data assignment and class addition on table 1`] = `
"
      <div id=\\"currency-table_body\\"><tr class=\\"currency-table_row\\"><td class=\\"currency-table_cell\\">gbpeur</td><td class=\\"currency-table_cell align-right\\">1.234401281999198</td><td class=\\"currency-table_cell align-right\\">1.255268054923685</td><td class=\\"currency-table_cell align-right\\">0.02224912511866406</td><td class=\\"currency-table_cell align-right\\">-0.010712388852019084</td><td class=\\"currency-table_cell align-right\\"><span></span></td></tr></div>
    "
`;

exports[`Sorted Datatable should receive message 1`] = `
"
      <div id=\\"currency-table_body\\"><tr class=\\"currency-table_row\\"><td class=\\"currency-table_cell\\">gbpeur</td><td class=\\"currency-table_cell align-right\\">1.234401281999198</td><td class=\\"currency-table_cell align-right\\">1.255268054923685</td><td class=\\"currency-table_cell align-right\\">0.02224912511866406</td><td class=\\"currency-table_cell align-right\\">-0.010712388852019084</td><td class=\\"currency-table_cell align-right\\"><span></span></td></tr></div>
    "
`;

exports[`Sorted Datatable should receive message, calculate midPrice, sort resultant data and render table 1`] = `
"
      <div id=\\"currency-table-body\\"><tr class=\\"currency-table-row\\"><td class=\\"currency-table-cell\\">usdjpy</td><td class=\\"currency-table-cell align-right\\">106.7297012204255</td><td class=\\"currency-table-cell align-right\\">107.25199883791178</td><td class=\\"currency-table-cell align-right\\">-2.8769211401569663</td><td class=\\"currency-table-cell align-right\\">-4.862314256927661</td><td class=\\"currency-table-cell align-right\\"><span></span></td></tr></div>
    "
`;

exports[`Sorted Datatable should receive message, data assignment and class addition on table 1`] = `
"
      <div id=\\"currency-table_body\\"><tr class=\\"currency-table_row\\"><td class=\\"currency-table_cell\\">gbpeur</td><td class=\\"currency-table_cell align-right\\">1.234401281999198</td><td class=\\"currency-table_cell align-right\\">1.255268054923685</td><td class=\\"currency-table_cell align-right\\">0.02224912511866406</td><td class=\\"currency-table_cell align-right\\">-0.010712388852019084</td><td class=\\"currency-table_cell align-right\\"><span></span></td></tr></div>
    "
`;
